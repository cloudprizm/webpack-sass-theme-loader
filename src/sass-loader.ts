import { RunnableTheme, resolvers } from '@hungry/bulma-theme'
import { traverseModuleRules, RuleSetRule, ruleSetUse, test } from '@hungry/webpack-parts'
import { ifElse, identity } from 'ramda'
import decamelize from 'decamelize'

const onlyPlainVars = (d: unknown) =>
  typeof d === 'string'
  || typeof d === 'number'
  || typeof d === 'boolean'

export const revertCamelCaseWithSizes = (str: string) =>
  decamelize(str, '-')
    .replace(/\_/g, '-')

export const toBulmaVariables = (theme: Record<string, any>): Record<string, unknown> =>
  Object
    .keys(theme)
    .filter(key => onlyPlainVars(theme[key]))
    .reduce((acc, camelCasedKey) => ({
      ...acc,
      [revertCamelCaseWithSizes(camelCasedKey)]: theme[camelCasedKey]
    }), {})

export const makeSassLoader = (theme: RunnableTheme) => ({
  loader: '@epegzz/sass-vars-loader',
  options: {
    syntax: 'sass',
    vars: theme.map(toBulmaVariables).run(resolvers)
  }
})

export const checkIfSass = ({ test: rule }: RuleSetRule) =>
  typeof rule === 'string'
    ? test.sass.test(rule)
    : rule instanceof RegExp
      ? rule.test('.sass')
      : false

export const ruleUseToArray = (rule: RuleSetRule['use']) =>
  typeof rule === 'string'
    ? [rule]
    : Array.isArray(rule)
      ? rule
      : []

export const injectSassVars = (theme: RunnableTheme) => ifElse(
  checkIfSass,
  ruleSetUse
    .modify(use => ruleUseToArray(use)
      .concat([makeSassLoader(theme)])),
  identity
)

export const injectSassVarLoader = (theme: RunnableTheme) =>
  traverseModuleRules
    .modify(injectSassVars(theme))

export const enableSassTheming = injectSassVarLoader