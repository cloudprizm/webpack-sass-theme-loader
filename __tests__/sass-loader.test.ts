import { checkIfSass, toBulmaVariables, enableSassTheming } from '../src/sass-loader'
import { theme } from '@hungry/bulma-theme'

test('check if sass', () => {
  const sassExtension = checkIfSass({ test: '.sass' })
  const sassRegexp = checkIfSass({ test: /\.s(a|c)ss$/ })

  expect(sassExtension).toBeTruthy()
  expect(sassRegexp).toBeTruthy()
})

test('convert to sass variables', () => {
  const plain = {
    black: '#000',
    hasSth: true,
    width: 10,
    size_1: 10
  }
  const lists = {
    colors: {
      black: '#0000'
    },
    sizes: [10]
  }
  const variables = toBulmaVariables({
    ...plain,
    ...lists
  })
  expect(Object.keys(variables)).toEqual(['black', 'has-sth', 'width', 'size-1'])

  expect(variables.colors).toBeUndefined()
  expect(variables.sizes).toBeUndefined()

  expect(variables).toMatchSnapshot()
})

test('embed sass loader on arbitrary webpack config', () => {
  const webpackConfig = {
    module: {
      rules: [{
        test: /\.sass$/,
        use: [
          'style-loader',
          'css-loader',
          'sass-loader'
        ]
      }]
    }
  }
  const withEmbededTheme = enableSassTheming(theme)(webpackConfig)
  // @ts-ignore
  const getLastChild = withEmbededTheme.module.rules[0].use[3]
  expect(getLastChild.options.vars).not.toBeNull()
})