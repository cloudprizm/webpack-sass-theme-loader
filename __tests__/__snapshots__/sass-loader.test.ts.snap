// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert to sass variables 1`] = `
Object {
  "black": "#000",
  "has-sth": true,
  "size-1": 10,
  "width": 10,
}
`;
